src/sentry/grouping/ingest/seer.py

```diff
@@ -0,0 +1,95 @@
+import logging
+from dataclasses import asdict
+
+from sentry import features
+from sentry.api.endpoints.group_similar_issues_embeddings import get_stacktrace_string
+from sentry.constants import PLACEHOLDER_EVENT_TITLES
+from sentry.eventstore.models import Event
+from sentry.grouping.grouping_info import get_grouping_info_from_variants
+from sentry.grouping.result import CalculatedHashes
+from sentry.models.group import Group
+from sentry.models.project import Project
+from sentry.seer.utils import (
+    SeerSimilarIssuesMetadata,
+    SimilarIssuesEmbeddingsRequest,
+    get_similarity_data_from_seer,
+)
+from sentry.utils.safe import get_path
+
+logger = logging.getLogger("sentry.events.grouping")
+
+
+def should_call_seer_for_grouping(event: Event, project: Project) -> bool:
+    """
+    Use event content, feature flags, rate limits, killswitches, seer health, etc. to determine
+    whether a call to Seer should be made.
+    """
+    # TODO: Implement rate limits, kill switches, other flags, etc
+    # TODO: Return False if the event has a custom fingerprint (check for both client- and server-side fingerprints)
+
+    if (
+        event.title in PLACEHOLDER_EVENT_TITLES
+        and not get_path(event.data, "exception", "values", -1, "stacktrace", "frames")
+        and not get_path(event.data, "threads", "values", -1, "stacktrace", "frames")
+    ):
+        return False
+
+    return features.has("projects:similarity-embeddings-metadata", project) or features.has(
+        "projects:similarity-embeddings-grouping", project
+    )
+
+
+def get_seer_similar_issues(
+    event: Event,
+    primary_hashes: CalculatedHashes,
+    num_neighbors: int = 1,
+) -> tuple[
+    dict[
+        str, str | list[dict[str, float | bool | int | str]]
```

Comment:
worth making a `TypedDict` for this type?

Reply:
_sigh_ This is why I say Python is the IHOP of typing. I agree this isn't great, but there's no way that I've been able to figure out (and googling suggests it's because no way exists) to create a typeddict out of a dataclass. 

So I can, but it'll just mean hard coding a second copy of all of the attributes from both the `SeerSimilarIssuesMetadata` and `SeerSimilarIssueData` dataclasses, which feels worse to me (I think because a change to either one of those classes would be guaranteed to make the hypothetical typeddict need updating whereas there's a good chance they could be changed without this type needing to be any different). 

WDYT?

Reply:
yeah totally. its good as is!

Reply:
Googling is so 2010s :-P Just ask Copilot ;-) 

```python
def dataclass_to_typeddict(cls):
    return TypedDict(cls.__name__ + 'Dict', {k: v for k, v in get_type_hints(cls).items()})

SeerSimilarIssuesMetadataDict = dataclass_to_typeddict(SeerSimilarIssuesMetadata)
SeerSimilarIssueDataDict = dataclass_to_typeddict(SeerSimilarIssueData)
```

Reply:
If only, @vartec. TIL about `get_type_hints`, but my very first attempt to wrestle with the dataclass/typeddict divide was to do something like this:

```python
fields = {"a": int, "b": str}
DerivedTypedDict = TypedDict("DerivedTypedDict", fields)
```

and then construct the dataclass from the same `fields` value. Alas, mypy will have none of it: `error: TypedDict() expects a dictionary literal as the second argument`.

Here's an open issue about it in the mypy repo: https://github.com/python/mypy/issues/4128.

Reply:
mypy is the worst...